Bearded.Utilities/Noise/PerlinNoise.cs

```diff
@@ -0,0 +1,113 @@
+        /// </param>
+        /// <param name="seed"></param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentOutOfRangeException"></exception>
+        public static IProceduralTexture Generate(int numCellsX, int numCellsY, int? seed)
+        {
+            if (numCellsX <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(numCellsX));
+            }
+            if (numCellsY <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(numCellsY));
+            }
+
+            var gradientArray = NoiseUtils.GenerateRandomArray(
+                // We generate the corners, but to make the noise map wrap along both dimensions, we reuse the same
+                // values as the left and top boundaries, so we don't have to generate the right and bottom boundaries.
+                numCellsX,
+                numCellsY,
+                r => vectorSamples[r.Next(vectorSamples.Length)],
+                seed);
+            return new PerlinProceduralTexture(gradientArray);
+        }
+
+        private sealed class PerlinProceduralTexture : IProceduralTexture
+        {
+            private readonly Vector2d[,] gradientArray;
+            private readonly int width;
+            private readonly int height;
+
+            public PerlinProceduralTexture(Vector2d[,] gradientArray)
+            {
+                this.gradientArray = gradientArray;
+                width = gradientArray.GetLength(0);
+                height = gradientArray.GetLength(1);
+            }
+
+            public double ValueAt(double x, double y)
+            {
+                if (x < 0 || x >= 1)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(x));
+                }
+                if (y < 0 || y >= 1)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(y));
+                }
+
+                // First we transform x and y from [0, 1) x [0, 1) to [0, width) x [0, height) for easier math.
+                x *= width;
+                y *= height;
+
+                // We take the floor and ceil to get the corners of the grid cell that surround the current point.
+                var xLower = (int) x;
+                var xUpper = xLower + 1;
+
+                var yLower = (int) y;
+                var yUpper = yLower + 1;
+
+                // Calculate dot products between distance and gradient for each of the grid corners
+                var topLeft = dotProductWithGridDirection(xLower, yUpper, x, y);
+                var topRight = dotProductWithGridDirection(xUpper, yUpper, x, y);
+                var bottomLeft = dotProductWithGridDirection(xLower, yLower, x, y);
+                var bottomRight = dotProductWithGridDirection(xUpper, yLower, x, y);
+
+                // Interpolation weights
+                var tx = Interpolation1.SmoothStep.Interpolate(0, 1, x - xLower);
```

Comment:
Are te interpolations used the only ones 'valid' for perlin noise?

Reply:
These are part of the definition of Perlin noise, yes.